동기화 메모장 [Step1] - Jost

YongHoonJJo opened the pull request:
@TaeJoongYoon 

타일러 안녕하세요! 개인적인 일도 좀 겹치고, STEP 2의 영역을 STEP 1 의 영역으로 찾각하여, PR 이 조금 늦어졌습니다. 
지난번 쥬스메이커 때, 타일러 덕분에 많은 것을 배울 수 있었고 생각할 수 있는 계기가 되었던 것 같았습니다. 
이번 플젝도 꼼꼼한 리뷰 부탁드리겠습니다!! 🙏

## < 동작화면 >

<table>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/13485924/132232923-dc43f8ce-5a7c-4174-bc45-929f362b345a.gif" /></td>
    <td><img src="https://user-images.githubusercontent.com/13485924/132233049-34853c55-2445-441c-b170-75022c59ffbe.gif" /></td>
  </tr>
</table>


## < 궁금한 점 >

### 오토레이아웃 코드로 구현하기.

- NSLayoutConstraint 방식의 경우 컴파일 타임에 오류 체크가 어려운 점이 있어, Layout Anchor 방식을 사용하게 되었습니다. Layout Anchor 의 경우 제약을 만들지 못하는 상황이 발생할 순 있다고 하는데, 프로젝트가 작아서인지, 아직은 경험이 부족하여 이점은 아직 잘 모르겠습니다. 타일러의 경우 오토레이아웃을 적용하게 될 경우 어떤 방식을 선호하는지 궁금합니다.



### 뷰컨트롤러, 테이블 뷰 커스텀 셀의 이니셜라이저

```swift
class MemoDetailViewController: UIViewController {
    private let memoTitle: String
    private let memoBody: String
    
    init(memo: Memo) {
        self.memoTitle = memo.title
        self.memoBody = memo.body
        
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
```

```swift
class MemoTableViewCell: UITableViewCell {
    static let identifier = "MemoListCell"
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        // ...
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}
```

- 우전 이전 쥬스메이커를 리뷰해주시면서 마지막에 위와 같이 뷰컨트롤러의 이니셜라이저를 통해 데이터를 전달할 수 있는 방법을 설명해주셔서, 그 부분을 항상 염두해두고 있다가 마침 적용할 기회가 생겨 적용해볼 수 있었습니다.!! 다만 코드에 fatalError() 를 넣는 것을 좋아하진 않는데, required init() 을 추가하지 않을경우 발생하는 에러메세지에서 Fix 를 하니 위와 같은 코드가 자동으로 생성되었습니다. 우선 위와 같이 작성하는 것이 적절한 방법인지 궁금하며, 적절하다면 fatalError() 대신 어떤식으로 처리하면 좋을지 궁금합니다.







### 메모 모델에서 스트링 쪼개기

```swift
private func generatePreviewText(from text: String) -> String {
  if text.count < Memo.maximumPreviewCount { return text }

  let previewEndIdx = text.index(text.startIndex,
                                 offsetBy: Memo.maximumPreviewCount)
  return text[text.startIndex...previewEndIdx].description
}
```

- 테이블 뷰의 셀에서 본문의 미리보기에 대한 일부 텍스트를 제공하는 요구사항이 있었습니다. 사실 위의 코드는 구글링하여 가져온 코드라, 혹시 스트링을 적당한 길이로 자르는 보다 좋은 로직이 있는지 궁금합니다. 
- 또한, 레이블의 경우 사이즈를 넘어가면 자동으로 ... 처리해주는 기능이 있는걸로 알지만, 텍스트의 양이 많아 적당히 잘라놓은 스트링으로 적용하는것이 성능상 막연히 생각하기에 이점이 있지 않을까 생각하여 이렇게 적용해봤습니다. 이런식으로 하는 것이 실제 성능적으로 유리한면이 있는지, 아니면 아무리 많은 텍스트여도 해당 텍스트를 그대로 활용하는 것이 더 나은 방법인지 궁금합니다.



### 모델 관련

```swift
struct Memo {
    static private let maximumPreviewCount = 50
    
    var title: String
    var body: String {
        willSet (newVal) {
            previewBody = generatePreviewText(from: newVal)
        }
    }
    private var lastModified: Double {
        willSet (newVal) {
            formatedLastModified = DateManager.transfromFormatedDate(from: newVal)
        }
    }
    var previewBody: String = ""
    var formatedLastModified: String?
    
    init(title: String, body: String, lastModified: Double) {
        self.title = title
        self.body = body
        self.lastModified = lastModified
        self.previewBody = generatePreviewText(from: body)
        self.formatedLastModified = DateManager.transfromFormatedDate(from: lastModified)
    }
}
```

- 셀을 구성하는데 필요한 포맷팅된 날짜의 경우도 위의 본문 미리보기 데이터와 마찬가지로 데이터를 미리 가지고 있는 방식을 선택하였습니다. 또한 미리보기 데이터와 날짜데이터는 원본 값에 의존하는 값들이라 willSet 의 기능을 이용해봤는데, 모델 파일이 좀 지저분해 보이는 감이 없잔아 있습니다. 필요한 프로퍼티에 대해서만 가지고 있는 방직으로 구현하는것이 적절한지, 이런식으로 구현하는것도 나쁘지 않은 선택인지 궁금합니다.



### 스플릿 뷰컨트롤러 구현 방식 관련

- 스플릿 뷰컨트롤러에 대한 자료가 많이 보이지 않아 구현하는 것이 쉽지 않았던 것 같습니다. 일단 워킹은 되는데, 맞게 이해를 한 상태로 구현을 한 것인지 모르겠습니다. ㅜ


<hr/>


### 이슈 1.

```
(
    "<NSAutoresizingMaskLayoutConstraint:0x6000033840f0 h=--& v=--& _UIButtonBarButton:0x7fb4a9570850.height == 0   (active)>",
    "<NSLayoutConstraint:0x6000033c36b0 _UIUCBKBSelectionBackground:0x7fb499504220.bottom == _UIButtonBarButton:0x7fb4a9570850.bottom - 6   (active)>",
    "<NSLayoutConstraint:0x6000033c3610 V:|-(6)-[_UIUCBKBSelectionBackground:0x7fb499504220]   (active, names: '|':_UIButtonBarButton:0x7fb4a9570850 )>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x6000033c36b0 _UIUCBKBSelectionBackground:0x7fb499504220.bottom == _UIButtonBarButton:0x7fb4a9570850.bottom - 6   (active)>
```

```swift
class MemoDetailViewController: UIViewController {
  private func updateLayout() {
        // ...
        let item = memoTextView.inputAssistantItem
        item.leadingBarButtonGroups = []
        item.trailingBarButtonGroups = []
    }
}
```

- 이 코드는 아이패드의 텍스트뷰 영역에서 입력모드가 활성화가 되면 아이패드에서 시스템 관련 메뉴가 올라오면서 위와 같은 워닝이 발생합니다. 위와 같이 코드를 추가하니 해당 오류는 해결 되었지만, 시스템 메뉴의 모양이 달라지긴 합니다. 이게 오토레이아웃 문제인지, 아니면 애플버그인지를 모르겠습니다. ㅜㅜ



### 이슈 2.

<img src="https://user-images.githubusercontent.com/13485924/132233136-62b5c596-b320-4c86-b9eb-b91ebb5277d3.gif" />


```swift
private func updateTextFontStyle() {
  // ...
  attributedStr.addAttribute(.font, value: titleFont,
                             range: (memoTextView.text as NSString).range(of: "\(memoTitle)\n"))
	// ...
}
```

- 타이틀 뒤의 개행을 안붙이면 타이틀에 공백이 없는 텍스트에 대해 커서가 폰트사이즈를 인식못하는 현상이 있었습니다. 일단 위와 같이 개행이 추가해서 해결은 했는데, 어떠한 이유에서 이런 현상이 나타나는질 몰겠습니다. ㅜ

<hr/>

### 그 외.

- 샘플 데이터를 파싱하는 코드는 샘플 코드를 위한 코드가 될 것 같다 별도의 스트링과 모델의 인스턴스를 생성하기 위한 데이더만 따로 가져와 하드코딩으로 작업해두었습니다. (나중에 삭제하기 용이하도록)
- 또한, Swift Lint 를 적용하면서 본문의 방대한 문자열 길이로 인해 최대 길이 제한이 걸려 별도의 파일로 분리하여, 린트가 적용되지 않도록 .swift.yml 파일에서 설정 하였습니다.
- 본격적인 CRUD 작업은 STEP 2 부터 진행하는 것으로 확인하여, 네이베이션 바의 + 버튼 터치시 우선 화면 이동만 되도록 구현해두었습니다.

YongHoonJJo committed:
etc: info.plist - Bundle Identifier 등록

YongHoonJJo committed:
feat: Memo 데이터 모델링.

- 데이터에 필요한 프로퍼티 선언
- 리스트시 필요한 미리보기용 데이터, 연산 프로퍼티 활용
- 샘플 데이터용 static 메서드 구현

YongHoonJJo committed:
refactor: initial viewcontroller 네이밍 수정 (MemoListViewController) 및 Controller 그룹으로 이동

YongHoonJJo committed:
etc: 프로젝트 Bundel Identifier 설정

YongHoonJJo committed:
feat: MemoList 에 대한 tableView 구현 (programmatically)

- DataSource 프로토콜 채택 및 준수.
- Layout Anchors 방식으로 테이블뷰 오토레이아웃 적용.

YongHoonJJo committed:
feat: MemoTableViewCell (커스텀 셀) 구현

- Layout Anchors 방식으로 contentView 내에 필요한 View 요소들 오토레이아웃 적용.
- 셀 악세사리(disclosureIndicator) 추가.
- configure() 메서드를 통한 Memo 데이터 바인딩

YongHoonJJo committed:
feat: 테이블 뷰 UITableViewDelegate 프로토콜 채택 및 didSelectRowAt 메서드 추가.

YongHoonJJo committed:
feat: MemoDetailViewController 추가

- UITextView 추가 및 오토레이아웃 적용
- 뷰컨트롤러 이니셜라이저를 통한 데이터 바인딩

YongHoonJJo committed:
refactor: 접근제어 추가 및 테이블 뷰 directionalMargin 으로 오토레이아웃 적용

YongHoonJJo committed:
feat: Memo Title 및 Body 에 따른 Font 적용

- 그 외 타이틀 및 본문 텍스트뷰 텍스트에 바인딩.
- 오토레이아웃 Bottom, Trailing Constant 정정

YongHoonJJo committed:
feat: UISplitVIewController 생성 및 계층구조 적용

- MemoList 를 Primary 로, MemoDetail 을 Secondary 로 적용
- 테이블 뷰에서 셀 탭할시 화면 전환 방식 변경 (네비게이션 컨트롤러에 감싼 후 showDetail 방식으로.)
- 메모 텍스트뷰에서 제목과 본문에 대한 seperator 변경 : 개행 두번으로.

YongHoonJJo committed:
feat: 화면 진입시 Primary View 머저 보여지도록 적용.

- 네비게이션 컨트롤러 타이틀 설정(매모)
- SizeClass 가 Regular 인 경우 네비게이션 컨트롤러에 이슈있음.

YongHoonJJo committed:
feat: iPad 텍스트뷰 터치시 키패드 관련 이슈로 인한 임시 해결 코드 적용.

YongHoonJJo committed:
feat: 테이블뷰 셀 디자인 적용

- Font Size 및 Color 적용
- 오토레이아웃 간격 조정

YongHoonJJo committed:
feat: 셀 터치 후 테이블 뷰 페이지로 돌아왔을 때 터치한 셀 배경색 원래 색상으로 되도록 적용.

YongHoonJJo committed:
feat: 음수처리의 경우 autoComplete 기능을 활용하기 위한 Int+Extension 적용

- nagative 라는 연산 프로퍼티 추가.

YongHoonJJo committed:
refactor: 셀 오토레이아웃 적용시 공통 값 상수를 통해 관리할 수 있도록 적용

YongHoonJJo committed:
feat: Date 포매팅을 위한 DateManager 타입 추가

- transfromFormatedDate(from:) timeInterval(Double)을 yyyy. MM. dd. 형식 변경하는 메서드 추가

YongHoonJJo committed:
feat: Memo 모델에 셀에 표시할 포맷팅된 날짜, 미리보기 문자열 관련 프로퍼티 추가.

- willSet 을 통한 body, lastModified 옵저빙.

YongHoonJJo committed:
feat: 디테일 뷰 이동시, 텍스트뷰 스크롤 Top 위치에서 시작되도록 적용.

YongHoonJJo committed:
feat: 스플릿 뷰 - preferredDisplayMode 변경

- .oneBesideSecondary 로 변경
- 아이폰 가로모드 및 아이패드 초기(세로모드) 에서 기본 방식 스플릿 안되는 현상, 위의 옵션으로 해결함.

YongHoonJJo committed:
fix: 타이틀 문자열에 공백이 없는 경우 커서가 Font-size 맞게 반응하지 않는 이슈 해결.

YongHoonJJo committed:
feat: 메모리스트 화면 상단 네비게이션 바에 BarButtonItem 추가

- 시스템 아이템 .add 로 적용
- action 에 대한 메서드 연결 (디테일 뷰 페이지로 이동 되는 것 까지만 적용)

YongHoonJJo committed:
etc: .gitignore 의존성관리 도구 관련 내용 추가

YongHoonJJo committed:
etc: pod install SwiftLint

YongHoonJJo committed:
etc: SwiftLint Build Phase 를 통한 스크립트 추가.

YongHoonJJo committed:
etc: SwiftLint 간단한 규칙 적용

- 라인 뒤 공백 허용
- 기본 파일 검사 제외
- 샘플 데이터용 	Memo+Extension.swift 파일 검사 제외

YongHoonJJo committed:
refactor: SwiftLint 룰에 맞게 수정.

TaeJoongYoon commented on CloudNotes/CloudNotes/Models/Memo.swift:
```diff
@@ -0,0 +1,43 @@
+            Memo(title: "네번째", body: "when I hear the buzz of the little world among the stalks, and grow familiar with the countless indescribable forms of the insects and flies, then I feel the presence of the Almighty, who formed us in his own image, and the breath\n\nI am so happy, my dear friend, so absorbed in the exquisite sense of mere tranquil existence, that I neglect my talents. I shou ...
+            Memo(title: "승리는 우리의 것", body: "I am alone, and feel the charm of existence in this spot, which was created for the bliss of souls like mine.\n\nI am so happy, my dear friend, so absorbed in the exquisite sense of mere tranquil existence, that I neglect my talents. I should be incapable of drawing a single stroke at the present moment; and yet I feel that I never was a greater artist  ...
+        ]
+    }
```
예시 json파일을 파싱하는게 아니라 내용만 가져오셨네요! 파싱하는게 프로젝트 목표라면 나중에 파싱하는 작업으로 바꿔도 좋겠네요 ㅎㅎㅎ

YongHoonJJo replied:
넵! 파싱하는 코드를 구현해도 샘플 데이터를 위한 코드가 되어 버려서, 또한 튜토리얼들을 보면 이런식으로 샘플데이터를 만들어서, 이런 선택을 하게 되었습니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/Models/Memo.swift:
```diff
@@ -0,0 +1,43 @@
+    var previewBody: String {
+        if body.count < Memo.maximumPreviewCount {
+            return body
+        }
```
이런 경우는 빠른 탈출을 위해 `guard`를 이용하는게 좋아보여요~

YongHoonJJo replied:
이게 최종 코드가 아닌데 이상하네요.ㅜ 수정 버전에서도 관련 내용은 동일하기 때문에 guard 로 변경하였습니다!! 이제는 guard 를 적절하게 잘 사용하고 있다고 생각하고 있었는데, 아직이었던 것 같습니다. ㅜ

TaeJoongYoon commented on CloudNotes/CloudNotes/Models/Memo.swift:
```diff
@@ -0,0 +1,43 @@
+            return body
+        }
+        
+        let previewEndIdx = body.index(body.startIndex,
```
`previewEndIdx`대신 `previewEndIndex`라고 적어주시면 더 좋을것 같네요~!

YongHoonJJo replied:
헐... 알고리즘 문제풀때의 습관이 아직 많이 남아있던 것 같습니다. 축약어 안쓰려고 했는데, 놓쳤던 것 같습니다. ㅜ 수정하였습니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/Models/Memo.swift:
```diff
@@ -0,0 +1,43 @@
+        let previewEndIdx = body.index(body.startIndex,
+                                       offsetBy: Memo.maximumPreviewCount)
+        return body[body.startIndex...previewEndIdx].description
+    }
```
본문을 preview에서 보여주기 위해 적당한 길이로 자르는 코드군요, 좋습니다 ㅎㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -7,12 +7,43 @@
+        setupTableView()
+    }
+    
+    func setupTableView() {
```
외부에서 사용할 일이 없다면 `private`을 붙여주세요 ㅎㅎㅎ

YongHoonJJo replied:
이부분도 최종 PR 전에 체크 후 private 을 붙였었는데 이상하네요 ㅜㅜ

TaeJoongYoon replied:
제가 최종코드가 아니라 커밋단위로 리뷰를 하다보니 생기는것 같네요 ㅎㅎ
나중에 보고 고쳐진 코드같으면 리뷰를 지우도록 해볼게요~

YongHoonJJo replied:
지우진 않으셔도 됩니다!! 커밋단위로 리뷰를 하신줄을 몰랐습니다. ㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -7,12 +7,43 @@
+        tableView.bottomAnchor.constraint(equalTo: safeArea.bottomAnchor).isActive = true
+        tableView.rightAnchor.constraint(equalTo: safeArea.rightAnchor).isActive = true
+        
+        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
```
cell의 `reuseIdentifier`는 상수로 관리해주시는게 실수를 줄이는 방법중 하나입니다 :)

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -7,12 +7,43 @@
 
 class MemoListViewController: UIViewController {
-
+    private let tableView = UITableView()
```
lazy하게 view 객체를 초기화해주는 방법도 있습니다 ㅎㅎ 그 방법도 한번 찾아보시면 좋을것 같아요 👍

YongHoonJJo replied:
```swift
lazy private var tableView: UITableView = {
    let tableView = UITableView()
    tableView.translatesAutoresizingMaskIntoConstraints = false
    tableView.register(MemoTableViewCell.self, forCellReuseIdentifier: MemoTableViewCell.identifier)
    return tableView
}()
```

이런식으로 View 를 생성하는 코드들이 있어 참고하고 변경해 보았습니다. 의도하신 바가 맞는지 궁금합니다!
또한, lazy 하게 view 객체를 초기화 해주는 것에는 어떠한 이점이 있는지 궁금했습니다.

If you are dealing with views (or anything else) that are always visible (needed) there is little point in using lazy initialization. On the other hand when you are dealing with instances that are not always needed - then using lazy var is justified.

요 내용이 가장 타당해 보였는데요, 이 프로젝트의 경우 위의 tableView 를 lazy 하게 초기화 해주는 것은 큰 이점은 없어 보입니다만, 프로젝트가 커지고, 위의 설명에서 나온 것 처럼 자주 다루지 않는 View 에 대해서 lazy 하게 초기화 하는 것에 메모리등의 이점이 있을 수 있다고 생각을 했습니다. 맞게 판단한것인지, 혹은 이 외 또다른 어떤 이점이 있는지 타일러의 의견이 궁급합니다.

추가로, 단점으로 thread-safe 하지 않다는 내용까지 알게 되었습니다! :)

TaeJoongYoon replied:
네네, view 객체 초기화 시에 한곳에서 볼수 있고 객체가 메모리에 lazy하게 올려지기 때문에 저는 주로 이런 방식으로 view를 초기화했었어요 ㅎㅎ
`thread-safe 하지 않다`는 아직까지 이 문제를 겪어보지 않아서 저도 좀더 찾아봐야겠군요.. 감사합니다 :)

YongHoonJJo replied:
그런 이점들도 있었군요!! ㅎㅎ 저도 감사합니다!! :)

TaeJoongYoon commented on CloudNotes/CloudNotes/Views/TableViewCells/MemoTableViewCell.swift:
```diff
@@ -0,0 +1,107 @@
+import UIKit
+
+class MemoTableViewCell: UITableViewCell {
+    static let identifier = "MemoListCell"
```
cell의 identifier를 상수로 관리하셨군요! 좋습니다 💯

TaeJoongYoon commented on CloudNotes/CloudNotes/Views/TableViewCells/MemoTableViewCell.swift:
```diff
@@ -0,0 +1,107 @@
+    func configure(on memo: Memo) {
+        memoTitleLabel.text = memo.title
+        memoTitleLabel.backgroundColor = .orange
+        lastModifiedLabel.text = memo.lastModified.description
```
`description`을 사용하신 이유가 있을까요??

YongHoonJJo replied:
이부분도 우선은 과거의 코드입니다. 또한 당시에 description 을 사용한 이유는 formatter 를 적용하기 전이었고, 그에 따라 Double 타입의  lastModified 값을 text 에 일단 넣고, 다른 내용들에 대해 확인 및 처리를 먼저하고자 함이었는데, 왜 리뷰를 하실 때 저 코드가 노출되었는지 모르겠습니다. ㅜ

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoDetailViewController.swift:
```diff
@@ -0,0 +1,56 @@
+    }
+    
+    private func updateLayout() {
+        let commonTextViewInset = 20
```
중복되는 inset값을 상수로 관리해주셨네요 ㅎㅎ 이러면 추후에 수정도 용이하고 실수를 줄일수 있어서 좋은 방법으로 보입니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoDetailViewController.swift:
```diff
@@ -26,12 +26,29 @@ class MemoDetailViewController: UIViewController {
+        attributedStr.addAttribute(.font, value: bodyFont, range: (memoTextView.text as NSString).range(of: memoBody))
+        
+        memoTextView.attributedText = attributedStr
+    }
```
이후 STEP의 작업을 미리 하셨군요 ㅎㅎㅎ
`NSMutableAttributedString`을 이용하여 한 `UITextView`안에 두 가지의 폰트를 보여주고 있네요. 💯 💯 
`addAttribute`도 있지만 `memoTitle`과 `memoBody`를 각각 `NSMutableAttributedString`와 `NSAttributedString`로 만들어 준 뒤 `append`하는 방법도 있습니다 ㅎㅎ

YongHoonJJo replied:
요구사항에 주어진 샘플 화면의 폰트 사이즈랑 굵기가 달라보여서 검색검색하면서 적용했는데, 알고보니 나중 스텝에서의 할 일 이었습니다. ㅜ

추가로 알려주신 내용에 대해서도 확인해보겠습니다. 감사합니다!!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/CloudMemoSplitViewController.swift:
```diff
@@ -0,0 +1,27 @@
+        let primaryNav = UINavigationController(rootViewController: primaryVC)
+        let secondaryNav = UINavigationController(rootViewController: secondaryVC)
+            
+        viewControllers = [primaryNav, secondaryNav]
```
변수명은 길어지더라도 `vc`->`viewController` / `nav`->`navigationController`로 사용해주시는게 좋습니다 :)

YongHoonJJo replied:
넵!! 수정하였습니다. 또한 이부분은 앞으로 플젝을 진행하면서 그렇게 진행하겠습니다. 감사합니다!!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/CloudMemoSplitViewController.swift:
```diff
@@ -0,0 +1,27 @@
+    
+    private func initContentHierarchy() {
+        let primaryVC = MemoListViewController()
+        let secondaryVC = MemoDetailViewController(memo: Memo.generateMemoList()[0])
```
`Memo.generateMemoList()[0]`가 만약 빈 배열이면 크래시가 날텐데 방어할 수 있는 코드가 필요해 보여요 :)

YongHoonJJo replied:
넵! 방어코드 추가하였습니다. 샘플데이터여서, 데이터가 존재하지 않을 가능성은 없을거라 판단하고 이렇게 했던 것 같습니다. ㅜ

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoDetailViewController.swift:
```diff
@@ -72,5 +72,9 @@ class MemoDetailViewController: UIViewController {
         
         view.backgroundColor = .white
+        
+        let item = memoTextView.inputAssistantItem
+        item.leadingBarButtonGroups = []
+        item.trailingBarButtonGroups = []
```
이런 코드는 별도의 함수로 관리해도 좋을것 같은데 어떠세요??? ㅎㅎ

YongHoonJJo replied:
여기에 작성하면서 따로 분리할까라고 망설인 기억이 있는데, 실행에 옮길껄 그랬었나봅니다 ㅜ 
별도로 분리하였습니다. 다만, 이 코드의 경우 해당 애플 버그(?)의 워닝을 막기 위한 용도라 분리한 메서드의 네이밍을 어떻게 해야할지 고민이 됩니다. 안에서 진행되는 코드들의 properties 를 기반으로 `clearBarButtonGroupsForInputAssistantItem()` 으로 명명했는데, 타일러는 이런 경우 어떤식으로 명명하는지 궁금합니다!

TaeJoongYoon replied:
명확한 네이밍이네요! 만약 저였어도 `clearInputAssistantItemButtons()` 이렇게 비슷하게 했을것 같아요 ㅎㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/Views/TableViewCells/MemoTableViewCell.swift:
```diff
@@ -66,7 +77,7 @@ extension MemoTableViewCell {
         contentView.addSubview(lastModifiedLabel)
         lastModifiedLabel.leadingAnchor.constraint(
```
예를들어 `lastModifiedLabel`이 `memoTitleLabel`의 leading을 맞춰야 한다면
`lastModifiedLabel.leadingAnchor.constraint(equalTo: memoTitleLabel.leadingAnchor).isActive = true`
로 사용해도 좋겠네요 ㅎㅎ

YongHoonJJo replied:
현재 플젝에서는 두 레이블의 Leading 이 동일하고, 제가 적용한 코드도 결국 두 레이블의 Leading 을 동일한 값으로 맞추기 위해 작성한 코드이기 때문에, 타일러가 제시해준 방법이 더 적합하다고 생각합니다. 수정하였습니다! :)

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -59,5 +59,11 @@ extension MemoListViewController: UITableViewDelegate {
         let detailNav = UINavigationController(rootViewController: memoDetailVC)
         self.showDetailViewController(detailNav, sender: self)
+        
+        guard let cell = tableView.cellForRow(at: indexPath) as? MemoTableViewCell else {
+            return
+        }
+        
+        cell.backgroundColor = .white
```
셀을 선택했을때, 배경색이 바뀌고 안돌아오는 이슈라면 아래 코드를 활용해보세요 :)
`tableView.deselectRow(at: indexPath, animated: true)`

YongHoonJJo replied:
헐!! 이런 방법이 있었다니!! 싶네요 ㅎㅎ 앞으로 잘 활용하겠습니다. 감사합니다!! :)

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoDetailViewController.swift:
```diff
@@ -68,7 +68,7 @@ class MemoDetailViewController: UIViewController {
         memoTextView.trailingAnchor.constraint(
             equalTo: safeArea.trailingAnchor,
-            constant: CGFloat(commonTextViewInset*(-1))
+            constant: CGFloat(commonTextViewInset.nagative)
```
`nagative`가 어떨때 사용되는지 궁금했는데 이런경우를 위해서 만드신 거군요 ㅎㅎㅎ
Swift에서는 단순 -를 붙여도 음수로 전환되기때문에 `-commonTextViewInset`로 넣으셔도 음수값으로 인식합니다!
이 부분은 한번더 고민해보시고 작업해보시면 좋을것 같아요~

YongHoonJJo replied:
만들 당시엔 나름 잘 만들었다고 생각했는데, PR 이후 다른 캠퍼들의 코드를 보고, 타일러가 제시한 방식으로도 가능하다는 것을 알게 되었습니다. 부끄럽네요. ㅜ 제시한 방법이 더 깔끔하고, 가독성 측면에서도 더 명확하다 생각합니다. 수정하겠습니다!!

TaeJoongYoon replied:
부끄러울게 어디있나요 :) 다 이유가 있어서 만들어진 코드니 괜찮습니다 ㅎㅎㅎ

YongHoonJJo replied:
익스텐션은 삭제하고 마이너스 연산자를 활용하였습니다!
다만 `-CGFloat(tailingConstant)` 와 `CGFloat(-tailingConstant)` 중 어느쪽이 나을지 고민됩니다. ㅜ

TaeJoongYoon replied:
음.. 저라면 후자처럼 작성할것 같아요! ㅎㅎ

YongHoonJJo replied:
그렇다면 저도..ㅎㅎ 감사합니다!!

TaeJoongYoon commented on CloudNotes/CloudNotes/Views/TableViewCells/MemoTableViewCell.swift:
```diff
@@ -45,39 +45,47 @@ extension MemoTableViewCell {
+        
+        updateTitleLayout(leadingConstant: cellLeadingConstant,
+                          verticalConstant: cellVerticalConstant,
+                          tailingConstant: cellTrailingConstant)
```
`tailingConstant`에서 `trailingConstant`으로 오타 수정이 필요해보여요!

YongHoonJJo replied:
헐.... 예리하십니다ㅜ 수정하였습니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/Managers/DateManager.swift:
```diff
@@ -0,0 +1,21 @@
+        let date = Date(timeIntervalSince1970: timeInterval)
+        return formatter.string(for: date)
+    }
+}
```
`DateManager`의 역할이 좀 더 Date에 맞춰서 확장된다면 좋을것 같네요.
지금 정도의 역할이라면 `Date`의 `extension`으로 만들어도 될것 같아요!

YongHoonJJo replied:
```swift
protocol Formattable {
    var formatter: DateFormatter { get }
}

extension Formattable {
    var formatter: DateFormatter {
        let dateFommatter = DateFormatter()
        dateFommatter.locale = Locale(identifier: "ko_KR")
        dateFommatter.dateStyle = .medium
        return dateFommatter
    }
}

extension Date: Formattable {
    func transfromFormattedType() -> String? {
        return formatter.string(for: self)
    }
}
```

Date의 extension 으로 만들어도 될 것 같다고 하셔서 위와 같이 변경해보았습니다. 적절하게 변경했는지, 혹은 생각하신 바와 좀 다른지 의견 부탁드리겠습니다!

TaeJoongYoon replied:
넵 ㅎㅎ Protocol과 extension까지 다양하게 사용하셨네요 ㅎㅎ 좋습니다 👍

TaeJoongYoon commented on CloudNotes/CloudNotes/Models/Memo.swift:
```diff
@@ -10,23 +10,34 @@ import Foundation
-                                       offsetBy: Memo.maximumPreviewCount)
-        return body[body.startIndex...previewEndIdx].description
     }
+    var previewBody: String = ""
+    var formatedLastModified: String?
```
이 두 변수를 이전처럼 연산 프로퍼티가 아닌 저장 프로퍼티로 관리하는 이유가 있을까요??

YongHoonJJo replied:
```swift
var body: String {
    willSet (newVal) {
        previewBody = generatePreviewText(from: newVal)
    }
}
private var lastModified: Double {
    willSet (newVal) {
        formatedLastModified = DateManager.transfromFormatedDate(from: newVal)
    }
}
var previewBody: String = ""
var formatedLastModified: String?
```

일단 두 프로퍼티는 body 와 lastModified 에 의존을 하는 프로퍼티라고만 생각해서 미리 값을 만들어 놓으면, 연산 속도에 이점이 있지 않을까 생각하고 만들었습니다. 하지만 다시 생각해보니, 오히려 Computed Property 를 활용하면, 연산속도는 이 정도 코드에서는 크게 차이는 없을 것 같고, 오히려 메모의 수가 많아질 수록 메모리를 차지하는 비용이 아껴질 수 있다는 것이 더 이점이 될 것 같단 생각이 들었습니다.

혹시 이런 부분을 의도하여 질문하신 것이 맞을까요?  코드는 Computed Property 로 수정하겠습니다!

TaeJoongYoon replied:
네네 말씀해주신 부분이 맞습니다! 값을 저장하고 이후 가공하는데 쓰이면 저장 프로퍼티로 하고, 단순 계산을 위한 프로퍼티면 연산 프로퍼티로 하는게 좋지 않을까 해서 코멘트 남겨봤어요 ㅎㅎ

YongHoonJJo replied:
변경하였고, 실제로 변경하니 코드도 더 깔끔해지는 마법(?)이 생겼습니다. ㅎㅎ

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/CloudMemoSplitViewController.swift:
```diff
@@ -35,3 +37,16 @@ extension CloudMemoSplitViewController: UISplitViewControllerDelegate {
+            return
+        }
+        
+        let memoDetailVC = MemoDetailViewController(memo: Memo.generateMemoList()[0])
+        let detailNav = UINavigationController(rootViewController: memoDetailVC)
+        primaryNaviVC.showDetailViewController(detailNav, sender: self)
```
실제 add에 대한 action도 디테일 뷰 페이지로 이동하는건가요???

TaeJoongYoon commented on CloudNotes/CloudNotes/Managers/DateManager.swift:
```diff
@@ -0,0 +1,21 @@
+
+class DateManager {
+    private static let formatter: DateFormatter = {
+       let f = DateFormatter()
```
`DateFormatter`에는 `locale`과 `dateStyle`속성이 있습니다. 이를 이용해보면 좋을것 같아요 👍 👍 

그리고 네이밍은 항상 신경써주세요 :)

YongHoonJJo replied:
`formatter.locale = Locale(identifier: "ko_KR")` 은 어느 나라 형식으로 날짜와 시간을 보여줄 지를 결정하고,
`dateStyle : .medium` 인 경우 → YYYY. MM. DD. 형태로 출력이 된다는 것을 확인하고 적용하였습니다!

네이밍 역시 최종 PR 전에는 변경하였지만, 이 부분은 첫 코드 작성시에도 신경 쓰도록 하겠습니다! :)

YongHoonJJo replied:
네이밍 역시 최종 PR 전에는 변경하였지만, 이 부분은 첫 코드 작성시에도 신경 쓰도록 하겠습니다! :)

TaeJoongYoon commented on CloudNotes/Podfile:
```diff
@@ -0,0 +1,19 @@
+target 'CloudNotes' do
+  # Comment the next line if you don't want to use dynamic frameworks
+  use_frameworks!
+  pod 'SwiftLint'
```
cocoapods까지.. 엄청난 선행학습이군요 👍

YongHoonJJo replied:
사실 처음에 적용하지 않은 이유는 `SwiftLint를 활용하여 코드 스타일을 일관성있게 유지합니다` 라는 제약사항이 있었는데 이를 보지 못해서였습니다...ㅎㅎ;
카르타고를 통해 설치를 해보려고 했는데, SwiftLint 는 카르타고를 통한 설치를 지원하지 않는 것 같아 CocoaPod 을 통해 설치를 하였습니다. 저희가 활동학습을 통해 CocoaPod, 카르타고, SPM 을 비교하는 시간을 가졌었는데요, 타일러의 현업에서는 어떤 의존성관리 도구를 사용하고 있으며, 개인적으로는 어떤 툴을 선호하시는지 궁금합니다!

TaeJoongYoon replied:
아하 이번 제약사항중에 그런게 있었군요 ㅎㅎㅎ
저는 주로 cocoapods을 사용했었는데 최근들어 개인프로젝트에서는 SPM을 사용해보려고 시도중이에요.
이유는 이전까지 안써본 도구기도 하고 애플에서 퍼스트파티로 제공해주는 모습이 보여서 입니다 ㅎㅎ

YongHoonJJo replied:
아아 그렇군요!! CocoaPod 을 이제 막 사용해본 상태지만, 추후 플젝에서는 사용 가능한 라이브러리가 SPM 에서 지원을 한다면, SPM 도 한번 시도해면 좋을 것 같네요 ㅎㅎ 의견 감사합니다!!

TaeJoongYoon commented on CloudNotes/.swiftlint.yml:
```diff
@@ -0,0 +1,10 @@
+disabled_rules: # Default Rules에서 비활성화할 규칙
+    # 라인 뒤에 공백이 없어야 합니다. https://realm.github.io/SwiftLint/trailing_whitespace.html
+    - trailing_whitespace
+
+
+excluded: # SwiftLint 검사에서 제외할 파일 경로
+  - Pods
+  - CloudNotes/AppDelegate.swift
+  - CloudNotes/SceneDelegate.swift
+  - CloudNotes/Extensions/Memo+Extension.swift
```
직접 추가한 규칙인가요??? ㅎㅎㅎ

YongHoonJJo replied:
네! 기본 규칙에는 메서드끼리 따닥따닥 붙이게 되어 있는데, 너무 제스타일이 아니라 확인해보고, 비활성화 시키는 방법이 있어서 ` trailing_whitespace` 를 추가하였습니다. 그리고 아직은 건드리지 않고 있는 기본 파일들에 대해서는 Lint 가 적용되진 않아도 될 것 같아서 검사에서 제외할 파일도 별도로 추가하였습니다!!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/CloudMemoSplitViewController.swift:
```diff
@@ -33,7 +33,9 @@ class CloudMemoSplitViewController: UISplitViewController {
 
 extension CloudMemoSplitViewController: UISplitViewControllerDelegate {
-    func splitViewController(_ splitViewController: UISplitViewController, collapseSecondary secondaryViewController: UIViewController, onto primaryViewController: UIViewController) -> Bool {
+    func splitViewController(_ splitViewController: UISplitViewController,
```
함수가 길어지는 경우, 첫 소괄호에서 개행을 해주시는 것도 좋습니다 ㅎㅎㅎ
[swift style guide](https://google.github.io/swift/#function-declarations)

YongHoonJJo replied:
```swift
public func index<Elements: Collection, Element>(
  of element: Element,
  in collection: Elements
) -> Elements.Index? where Elements.Element == Element, Element: Equatable {
  for current in elements {
    // ...
  }
}
```

위의 스타일도 좋은 것 같습니다. 적용하도록 하겠습니다. 좋은 참고 자료 감사합니다!!

TaeJoongYoon reviewed (changes requested):
안녕하세요 Jost! 타일러입니다 :)
Step을 착각하셔서 조금 늦어지셨군요..ㅜ 그래도 많은 부분 작업하시고 선행작업 하셨으니 한편으론 좋게 생각해도 될거같아요 ㅎㅎㅎ
이번에도 많은도움 드릴수 있도록 노력해볼게요.. ㅎㅎ

---
## 1. 오토레이아웃 코드로 구현

저같은 경우는 대부분 `NSLayoutConstraint.active()`로 구현을 합니다. 아직까지 컴파일타임에 오류체크를 못하는 경우는 별로 없어서 크게 생각해본 이슈는 아니네요..! 하지만 런타임중에 오토레이아웃이 변경될 경우가 있다면 그 때는 `NSLayoutConstraint`를 변수로 받아서 값을 조정하고는 합니다 ㅎㅎ

## 2. 뷰컨트롤러, 테이블 뷰 커스텀 셀의 이니셜라이저

우선 이니셜라이저를 잘 만들어 주신것 같아요. 해당 뷰 컨트롤러나 셀에서 필요한 정보를 생성자에서 받아서 잘 처리해주신것 같습니다 ㅎㅎ
`required init?(coder: NSCoder)`이 생기는 이유에 대해서는 **필수 이니셜라이저**(required initialzing)에 대해서 더 찾아보시면 도움이 될듯합니다!
`fatalError`로 그대로 둬도 되지만 그게 싫다면 단순히 `self.init()`을 호출해도 동작합니다 :)

## 3. 메모 모델에서 스트링 쪼개기

지금과 같이 index를 이용해서 자르시면 될것 같아요. 하지만 `description`을 사용하신 이유가 궁금하긴 하네요 ㅎㅎㅎ
`String()`으로 감싸줘도 올바른 값이 반환되긴 할겁니다!!
말씀해주신것처럼 `UILabel`에서는 말줄임이 되기 때문에 굳이 text를 자르지 않아도 표현하는데 있어서는 큰 문제는 없을거에요.
저도 테스트해보지 않아서 어느면이 성능적으로 유리한지는 모르겠지만 이런 고민을 하시고 작업을 하신 것에 큰 감동을 느끼네요.
따로 이런코드를 작성하지 않아도 정상적으로 동작하기 때문에 성능에 대한 고민을 안하고 바로 작업해도 될 부분인데
이렇게 고민을 남겨주시고 작업을 해주셔서 감사합니다. 어느 면이 성능적으로 나을지는 저도 고민해볼게요! 💯 💯 💯 

## 4. 모델 관련

위의 코멘트에도 적어놨지만 현재 `Memo`모델에서 `previewBody`와 `formatedLastModified`가 저장 프로퍼티로 존재해야 하는 이유가 궁금했어요. 이 두 프로퍼티는 값을 저장해서 사용하는게 아닌 각각 `body`와 `lastModified`에 의존하여 값이 정해지는 프로퍼티이고 읽음 전용이기 때문에 연산 프로퍼티로도 관리할 수 있어보이는데요. 만약 의도적으로 저장 프로퍼티로 관리하신 거라면 지금처럼 `willSet`이나 `didSet`을 이용하는 방법이 좋아보이네요. 👍 

## 5. 스플릿 뷰컨트롤러 구현 방식 관련

딱히 이상이 있어보이지는 않습니다 ㅎㅎ 요구사항에 맞게 잘 구현해주신것 같아요 :)

## 6. 아이패드 UITextView autolayout 이슈

음.. 저도 찾아보니 이건 애플버그같네요. 아래 포럼글에서도 그렇고 말씀해주신 아이패드에서 text를 입력/수정하는 상황에서 발생하는 워닝같습니다.
[apple forum](https://developer.apple.com/forums/thread/667441)

## 7. 개행문자 폰트사이즈 인식

`memoTitle`의 마지막 문자가 개행문자(\n)임에도 불구하고 초기 코드는 `memoTitle`까지만 font attribute가 적용되어 폰트사이즈를 인식 못하고 있는것 같네요. 그래서 이후 수정코드 혹은 아래코드 처럼 개행문자가 들어갈 경우, 개행문자까지 속성을 적용시켜줘야 커서가 폰트사이즈를 정확히 인식하는것 같습니다. ㅎㅎ

```swift
attributedStr.addAttribute(.font, value: titleFont, range: (memoTextView.text as NSString).range(of: "\(memoTitle)"))
attributedStr.addAttribute(.font, value: titleFont, range: (memoTextView.text as NSString).range(of: "\n"))
```

---

이번 STEP도 고생많으셨습니다. cocoapods도 사용해보시고 swiftLint도 사용해보시고 많은 시도를 해보시는군요 ㅎㅎ
계속해서 코드에도 고민의 흔적이 남아있는것 같아 좋습니다..!
피드백을 한번 살펴보신 후에 논의 및 적용을 해보시죠 :)

YongHoonJJo committed:
refactor: Memo.swift 피드백 반영

- Early Exit 적용을 위한 if 문을 guard 문으로 대체.
- 네이밍 수정 : previewEndIdx 에서 previewEndIndex 로 변경

YongHoonJJo committed:
refactor: 피드백 반영

- 네이밍 수정 : VC 에서 ViewController 로 변경
- 네이밍 수정 : Nav 에서 NavigationController 로 변경

YongHoonJJo committed:
refactor: 아이패드에서의 키패드 이슈를 해결하기 위한 코드 - 메서드로 분리

YongHoonJJo committed:
refactor: 테이블 뷰 lazy 하게 초기화 하는 코드로 변경

YongHoonJJo committed:
refactor: 피드백 반영

- updateLasModified 레이블에 대해 titleLabel 의 Leading 과 일치시키면 됨으로 인해 리딩에 상수를 적용하는 대신 titleLabel 의 리딩에 맞추도록 적용.

YongHoonJJo committed:
refactor: 첫번째 메모 데이터 바인딩시 예외처리 적용

- 추가로 코드가 긴 메서드들에 대한 swift style guide 에 의한 인덴트 적용.
- 뷰컨트롤러, 네비게이션 뷰컨트롤러 관련 네이밍 풀네임으로 변경.

YongHoonJJo committed:
refactor: 셀 선택 후 선택 배경색  되돌리는 기능 관련 코드 변경

-  tableView.deselectRow(at:animated:)  으로 적용

YongHoonJJo committed:
refactor: 프로퍼티 종류 변경, 타입 수정

- previewBody, formatedLastModified 저장 프로퍼티 대신 연산 프로퍼티 활용
- Double 대신 TimeInterval 로 변경

YongHoonJJo committed:
refactor: 음수값 마이너스 연산자를 통해 적용

- Int+Extension 파일 불필요해짐으로 삭제.

YongHoonJJo committed:
refactor: DateFormatter에는 locale과 dateStyle 속성을 활용한 날짜 출력 적용

YongHoonJJo committed:
refactor : Date의 extension 기능을 활용하여 formmater 기능 적용

- 기존의 DateManager.swift 삭제
- 새로운 역할의 Date+Extension.swift 추가.

YongHoonJJo commented:
### 1. 오토레이아웃 코드로 구현

- 경험 공유해주셔서 감사합니다. 저도 `NSLayoutConstraint` 오 같이 공부해보겠습니다!

### 2. 뷰컨트롤러, 테이블 뷰 커스텀 셀의 이니셜라이저

- Swift 는 이니셜라이저가 조금 어려운 언어인 것 같습니다. ㅜ required 에 대해서는 조금 더 공부해보겠습니다!

### 3.메모 모델에서 스트링 쪼개기

- description 사용 이유에 대해서는 리뷰에 답글 남겼습니다만..!! 애초에 올바른 방법으로 진행하겠습니다. ㅎㅎ
- 텍스트 뷰에 텍스트 양은 어마어마해질 수 있다보니 고민했던 내용이었는데, UILabel 의 lineBreak Mode 의 .byTruncatingTail 방식에 따른 내부 구현은 모르겠지만, 스트링의 길이 정도는 알 수 있으니 알아서 적당히 잘라서 계산해서 그렇게 표현하지 않을까란 생각도 드네요. 말씀하신 것 처럼 텍스트 양에 상관 없이 굳이 자르지 않고 사용해도 될 것 같단 생각이 듭니다 !

### 4. 모델 관련

- 리뷰 반영하여 수정하였습니다. :)

### 5. 스플릿 뷰컨트롤러 구현 방식 관련

- 감사합니다! 다만, 리펙토링 과정에서 셀을 터치했을 때, 즉 didSelectRowAt 가 호출되는 메서드에서 리스트 뷰컨트롤러가 `self.showDetailViewController(secondaryNavigationController, sender: self)` 와 같이 showDetail 를 직접 호출을 하는데, 이부분은 스플릿뷰를 리스트뷰의 델리게잇으로 두어 `        delegate?.showDetailViewController( memoList[indexPath.row])` 와 같이 호출하는 것이 더 바람직하지 않을까 하는 내용으로 다른 캠퍼와 의견을 주고 받았었습니다. 

- All view controllers implement this method. If they do not show the view controller themselves, they forward the message to the parent split view controller (if any) and let it show the view controller. As a result, a child view controller can call this method on itself to achieve the same results as calling this method on the split view controller object.

  - https://developer.apple.com/documentation/uikit/uisplitviewcontroller/1623182-showdetailviewcontroller

- 공식문서에서는 위와 같이도 표현을 하고 있어서 저의 선택이 나쁘진 않았다고 생각은 하는데, 이부분에 대해서는 어떻게 생각하시는지 궁금합니다! 

```swift
  extension MemoListViewController: UITableViewDelegate {
      func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
          tableView.deselectRow(at: indexPath, animated: true)
          
          let memoDetailViewController = MemoDetailViewController(memo: memoList[indexPath.row])
          let secondaryNavigationController = UINavigationController(rootViewController: memoDetailViewController)
          self.showDetailViewController(secondaryNavigationController, sender: self)
      }
  }
```

```swift
  extension MemoListViewController: UITableViewDelegate {
    weak var delegate: ShowDetailDelegate?
  	func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
          tableView.deselectRow(at: indexPath, animated: true)
          
          delegate?.showDetailViewController( memoList[indexPath.row])
      }
  }
  
  extension CloudMemoSplitViewController {
    
    let primaryViewController = MemoListViewController()
    primaryViewController.delegate = self
  
    // ...
    
    func showDetailViewController(_ memo: Memo) { // delegate Method
        guard let primaryNavigationController = viewControllers.first else { return }
          
        let memoDetailViewController = MemoDetailViewController(memo: memo)
        let secondaryNavigationController = UINavigationController(rootViewController: memoDetailViewController)
        primaryNavigationController.showDetailViewController(secondaryNavigationController, sender: self)
    }
}
```

- 요약하면 위 vs 아래의 코드가 됩니다..!



### 6. 아이패드 UITextView autolayout 이슈

- 애플 버그..ㅜ
- 같은 포럼을 보셨던 것 같습니다. ㅎㅎ



### 7. 개행문자 폰트사이즈 인식

 마지막 문자가 개행문자여서 그랬던 거군요!! 생각 못했었는데 감사합니다!!



### 그 외.

- 샘플데이터를 가져오는 코드에 가드문을 적용하면서 스플릿 뷰컨에 아래 두개의 메서드를 추가하면서 적용을 하였습니다. 그런데 이 코드들의 위치를 아무리봐도 이곳에 있으면 안될 것 같단 생각이 들었습니다. 어떻게 분리하는 것이 좋을지 계속 고민은 해봤는데, 이후에 코어데이터를 적용하면 또 어떻게 설계가 되어야 할까 고민하다가 시간이 후쩍 흘러버렸습니다. 이부분은 이 후 스텝에서 같이 진행행해봐도 괜찮을까요?? 위의 델리게인 관련 논쟁과 이부분에 대해 고민하드라 리뷰에 대한 피드백이 늦어졌습니다. ㅜ

```swift
class CloudMemoSplitViewController {
    private func generateNewMemo() -> Memo {
        let currentTime = Date().timeIntervalSince1970
        return Memo(title: "새로운 메모", body: "추가 텍스트 없음", lastModified: currentTime)
    }
    
    private func getFirstMemo() -> Memo {
        guard let memo = Memo.generateMemoList().first else {
            return generateNewMemo()
        }
        
        return memo
    }
}
```

YongHoonJJo committed:
refactor: 네이밍 수정

- memoDetailVC 를 memoDetailViewController 로.
- detailNav 를 secondaryNavigationController 로 변경.

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/MemoListViewController.swift:
```diff
@@ -7,7 +7,12 @@
 
 class MemoListViewController: UIViewController {
-    private let tableView = UITableView()
+    lazy private var tableView: UITableView = {
```
접근제어자인 `private`을 우선적으로 작성해주시는게 좋습니다!

YongHoonJJo replied:
어이쿠, 순서가 아직 익숙하지 않았던 것 같습니다. 이후 스텝에 반영 토록 하겠습니다!

TaeJoongYoon commented on CloudNotes/CloudNotes/Controllers/CloudMemoSplitViewController.swift:
```diff
@@ -16,39 +16,54 @@ class CloudMemoSplitViewController: UISplitViewController {
         self.delegate = self
     }
+    
+    private func generateNewMemo() -> Memo {
+        let currentTime = Date().timeIntervalSince1970
+        return Memo(title: "새로운 메모", body: "추가 텍스트 없음", lastModified: currentTime)
+    }
+    
+    private func getFirstMemo() -> Memo {
+        guard let memo = Memo.generateMemoList().first else {
+            return generateNewMemo()
+        }
+        
+        return memo
+    }
```
이 `Memo`를 반환하는 함수들은 뷰의 역할인`CloudMemoSplitViewController`보다 현재로써 모델의 역할을 하는 `Memo`에 있는게 나아보여요 ㅎㅎㅎ

TaeJoongYoon replied:
그리고 함수가 아닌 연산 프로퍼티로도 작성가능해 보이는군요 :)

YongHoonJJo replied:
넵 !! 저도 이 메서드들이 있을 곳이 여기는 아닌것 같았습니다. ㅜ 이 후 스텝에서 반영하도록 하겠습니다!

TaeJoongYoon reviewed (approved):
안녕하세요 Jost~ 타일러입니다 :)

이번 STEP1 잘 마무리 해주셔서 감사합니다 ㅎㅎ 많은 피드백양에도 불구하고 빠르게 적용해주셨네요.. 대단하십니다..! 👍 
피드백 반영된 커밋들은 다 확인했고 남겨주신 내용중을 보면 될것 같고 STEP2도 잘 부탁드릴게요 :)

---
## 1. showDetailViewController

이건 Jost 및 다른 캠퍼분들 의견대로 상위 뷰컨인 `CloudMemoSplitViewController`에서 하는게 더 직관적이겠네요 ㅎㅎㅎ
대신 delegate 변수 네이밍을 좀 더 명확히 하면 헷갈릴 여지가 줄어들것 같아요! 좋은 방법인듯 합니다 ㅎㅎ

## 2 샘플데이터를 가져오는 코드

위 코멘트에도 적어놨지만 이런 데이터 영역은 뷰보단 모델, 컨트롤러에서 담당하는게 적합해 보여요.  `CloudMemoSplitViewController`에 대한 뷰모델/컨트롤러를 따로 만들어서 관리 혹은 데이터를 관장하는 객체를 설계하는것이 좋아보입니다 ㅎㅎ

---

YongHoonJJo commented:
@TaeJoongYoon 

주말임에도 이렇게 리뷰 달아주셔서 감사합니다!!

### 1. showDetailViewController
- 의견 반영하여 다음 스텝에 적용해보도록 하겠습니다! 네이밍도 고민해보겠습니다. ㅜ

### 2. 샘플데이터를 가져오는 코드
- 이 부분은 데이터를 관리하는 Manager 를 따로 두어야 할 지, 그런 점이 고민이긴 했었습니다. ㅜ 잘 고민해보고 이 후 스텝에 적용해보겠습니다!! :)
